src/zmalloc.c

```diff
@@ -67,10 +68,34 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define MAX_THREADS 16 /* Keep it a power of 2 so we can use '&' instead of '%'. */
+#define THREAD_MASK (MAX_THREADS - 1)
+
+typedef struct used_memory_entry {
+    redisAtomic long long used_memory;
+    char padding[CACHE_LINE_SIZE - sizeof(long long)];
+} used_memory_entry;
 
-static redisAtomic size_t used_memory = 0;
+static __attribute__((aligned(CACHE_LINE_SIZE))) used_memory_entry used_memory[MAX_THREADS];
+static redisAtomic size_t num_active_threads = 0;
+static __thread long my_thread_index = -1;
+
+static inline void init_my_thread_index(void) {
+    if (unlikely(my_thread_index == -1)) {
+        atomicGetIncr(num_active_threads, my_thread_index, 1);
+        my_thread_index &= THREAD_MASK;
+    }
+}
+
+static void update_zmalloc_stat_alloc(long long num) {
+    init_my_thread_index();
+    atomicIncr(used_memory[my_thread_index].used_memory, num);
```

Comment:
Looks like the atomic is not necessary here, the thread already has its own slot, only one writer exists.
Ref: https://github.com/valkey-io/valkey/pull/674

Reply:
There might be more than one writer, we reuse slots if we have more than 16 threads.